File: 剑指 Offer/剑指 Offer 57 - II. 和为s的连续正数序列/cpp/main3.cpp
class Solution {
public:
    vector<vector<int>> findContinuousSequence(int target) {
    	vector<vector<int>> res;
    	int left = 1, right = 2;
        while (left < right) {
            int sum = (left + right) * (right - left + 1) / 2;
            if (sum == target) {
                vector<int> temp;
                for (int i = left; i <= right; ++i) {
                    temp.push_back(i);
                }
                res.push_back(temp);
                ++left;
            }
            else if (sum < target) {
                ++right;
            }
            else {
                ++left;
            }
        }
    	return res;
    }
};

File: codes/0653-Two-Sum-IV - Input-is-a-BST/Java/main3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public boolean findTarget(TreeNode root, int k) {
        if (root == null) {
            return false;
        }
        List<Integer> list = new ArrayList<>();
        // 使用中序遍历，得到一个顺序数组
        inOrder(root, list);
        int len = list.size();
        int l = 0;
        int r = len - 1;

        // 在顺序数组中，使用二分查找法
        while (l < r) {
            int sum = list.get(l) + list.get(r);

            if (sum > k) {
                r--;
            } else if (sum < k) {
                l++;
            } else {
                assert sum == k;
                return true;
            }
        }
        return false;
    }

    private void inOrder(TreeNode node, List<Integer> list) {
        if (node == null) {
            return;
        }
        inOrder(node.left, list);
        list.add(node.val);
        inOrder(node.right, list);
    }
}